protocol/shannon/websocket_context.go

```diff
@@ -0,0 +1,184 @@
+package shannon
+
+import (
+	"fmt"
+	"net/http"
+	"strconv"
+	"time"
+
+	"github.com/buildwithgrove/path/observation"
+	protocolobservations "github.com/buildwithgrove/path/observation/protocol"
+	"github.com/buildwithgrove/path/request"
+	"github.com/pokt-network/poktroll/pkg/relayer/proxy"
+	servicetypes "github.com/pokt-network/poktroll/x/service/types"
+	sessiontypes "github.com/pokt-network/poktroll/x/session/types"
+	sharedtypes "github.com/pokt-network/poktroll/x/shared/types"
+	sdk "github.com/pokt-network/shannon-sdk"
+)
+
+// GetWebsocketConnectionHeaders returns headers for the websocket connection:
+//   - Target-Service-Id: The service ID of the target service
+//   - App-Address: The address of the session's application
+//   - Rpc-Type: Always "websocket" for websocket connection requests
+func (rc *requestContext) GetWebsocketConnectionHeaders() http.Header {
+	headers := http.Header{}
+
+	// If the selected endpoint is a protocol endpoint, add the headers
+	// that the RelayMiner requires to forward the request to the Endpoint.
+	//
+	// Requests to fallback endpoints bypass the protocol so RelayMiner headers are not needed.
+	if !rc.getSelectedEndpoint().IsFallback() {
+		headers = rc.getRelayMinerConnectionHeaders(rc.getSelectedEndpoint().Session().GetHeader())
+	}
+
+	return headers
+}
+
+// getRelayMinerConnectionHeaders returns headers for RelayMiner websocket connections:
+//   - Target-Service-Id: The service ID of the target service
+//   - App-Address: The address of the session's application
+//   - Rpc-Type: Always "websocket" for websocket connection requests
+func (rc *requestContext) getRelayMinerConnectionHeaders(sessionHeader *sessiontypes.SessionHeader) http.Header {
+	if sessionHeader == nil {
+		rc.logger.Error().Msg("❌ SHOULD NEVER HAPPEN: Error getting relay miner connection headers: session header is nil")
+		return http.Header{}
+	}
+
+	return http.Header{
+		request.HTTPHeaderTargetServiceID: {sessionHeader.ServiceId},
+		request.HTTPHeaderAppAddress:      {sessionHeader.ApplicationAddress},
+		proxy.RPCTypeHeader:               {strconv.Itoa(int(sharedtypes.RPCType_WEBSOCKET))},
+	}
+}
+
+// GetWebsocketEndpointURL returns the websocket URL for the selected endpoint.
+// This URL is used to establish the websocket connection to the endpoint.
+func (rc *requestContext) GetWebsocketEndpointURL() (string, error) {
+	websocketURL, err := rc.selectedEndpoint.WebsocketURL()
+	if err != nil {
+		rc.logger.Error().Err(err).Msg("❌ Selected endpoint does not support websocket RPC type")
+		return "", err
+	}
+
+	return websocketURL, nil
+}
+
+func (rc *requestContext) SignClientWebsocketMessage(msgData []byte) ([]byte, error) {
```

Comment:
This is an internal detail of protocol, please consider refactoring to encapsulate this with a private method.